src/process.c

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#define _GNU_SOURCE
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+#if defined _WIN32 || defined __CYGWIN__
+#include <Windows.h>
+#else
+#include <libgen.h>
+#include <unistd.h>
+#endif
+
+#include "rcutils/allocator.h"
+#include "rcutils/error_handling.h"
+#include "rcutils/process.h"
+
+int rcutils_get_pid(void)
+{
+#if defined _WIN32 || defined __CYGWIN__
+  return (int)GetCurrentProcessId();
+#else
+  return (int)getpid();
+#endif
+}
+
+char * rcutils_get_executable_name(rcutils_allocator_t allocator)
+{
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
+    &allocator, "invalid allocator", return NULL);
+
+#if defined __APPLE__
+  const char * appname = getprogname();
+#elif defined __GNUC__
+  const char * appname = program_invocation_name;
+#elif defined _WIN32 || defined __CYGWIN
+  char appname[MAX_PATH];
+  int32_t size = GetModuleFileNameA(NULL, appname, MAX_PATH);
+  if (size == 0) {
+    return NULL;
+  }
+#else
+#error "Unsupported OS"
```

Comment:
I haven't found examples of other cross-platform code where we explicitly error in fall-through cases. I don't know if that's because other code is structured to assume Linux/Unix in the fall-through case or if runtime errors are being used instead.

Reply:
I had more to say about this, but GitHub deleted my review... I'm fine with it as-is.

Reply:
> I had more to say about this, but GitHub deleted my review...

:rage4: